service/src/main/java/greencity/service/HabitAssignServiceImpl.java

```diff
@@ -412,4 +424,62 @@ private int countNewHabitStreak(List<HabitStatusCalendar> habitCalendars) {
             habitAssign -> buildHabitAssignDto(habitAssign, language)).collect(Collectors.toList());
     }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public List<HabitsDateEnrollmentDto> findActiveHabitAssignsBetweenDates(Long userId, LocalDate from, LocalDate to,
+        String language) {
+        List<HabitAssign> list = habitAssignRepo.findAllActiveHabitAssignsBetweenDates(userId, from, to);
+        List<LocalDate> dates = Stream.iterate(from, date -> date.plusDays(1))
+            .limit(ChronoUnit.DAYS.between(from, to.plusDays(1)))
+            .collect(Collectors.toList());
+        List<HabitsDateEnrollmentDto> dtos = new ArrayList<>();
+        for (LocalDate date : dates) {
+            dtos.add(HabitsDateEnrollmentDto.builder().enrollDate(date)
+                .habitAssigns(new ArrayList<>())
+                .build());
+        }
+        list.forEach(habitAssign -> buildHabitsDateEnrollmentDto(habitAssign, language, dtos));
+
+        return dtos;
+    }
+
+    /**
+     * Method to fill in all user enrollment activity in the list of
+     * {@code HabitsDateEnrollmentDto}'s by {@code HabitAssign}'s list of habit
+     * status calendar.
+     *
+     * @param habitAssign {@code HabitAssign} habit assign.
+     * @param language    {@link String} of language code value.
+     * @param list        of {@link HabitsDateEnrollmentDto}.
+     */
+    private void buildHabitsDateEnrollmentDto(HabitAssign habitAssign, String language,
+        List<HabitsDateEnrollmentDto> list) {
+        HabitTranslation habitTranslation = getHabitTranslation(habitAssign, language);
+
+        for (HabitsDateEnrollmentDto dto : list) {
+            if (dto.getEnrollDate()
+                .isBefore(habitAssign.getCreateDate().toLocalDate().plusDays(habitAssign.getDuration() + 1L))
+                && dto.getEnrollDate()
+                    .isAfter(habitAssign.getCreateDate().toLocalDate().minusDays(1L))) {
+                boolean isDateEnrolled = false;
+                for (int i = 0; i < habitAssign.getHabitStatusCalendars().size(); i++) {
+                    if (habitAssign.getHabitStatusCalendars().get(i).getEnrollDate()
+                        .equals(dto.getEnrollDate())) {
+                        dto.getHabitAssigns().add(HabitEnrollDto.builder()
+                            .habitDescription(habitTranslation.getDescription()).habitName(habitTranslation.getName())
+                            .isEnrolled(true).habitId(habitAssign.getHabit().getId()).build());
+                        isDateEnrolled = true;
+                    }
+                }
+                if (!isDateEnrolled) {
+                    dto.getHabitAssigns().add(HabitEnrollDto.builder()
+                        .habitDescription(habitTranslation.getDescription()).habitName(habitTranslation.getName())
+                        .isEnrolled(false).habitId(habitAssign.getHabit().getId()).build());
+                }
+            }
+        }
```

Comment:
Don't understand what is being done here. Split to several private methods.

Reply:
Thanks, changed